### PR TITLE
Move fingerprinters to the core project

### PR DESCRIPTION
**Pull request by wolfs**
So that the snapshots project doesn't depend on `FileCollection` at all.

### CODE REVIEW

File: subprojects/execution/src/integTest/groovy/org/gradle/internal/execution/IncrementalExecutionIntegrationTest.groovy [hunk 1]
```diff
@@ -79,7 +80,8 @@ class IncrementalExecutionIntegrationTest extends Specification {
 
     def fileHasher = new TestFileHasher()
     def fileSystemMirror = new DefaultFileSystemMirror(Stub(WellKnownFileLocations))
-    def snapshotter = new DefaultFileSystemSnapshotter(fileHasher, new StringInterner(), TestFiles.fileSystem(), fileSystemMirror)
+    def fsSnapshotter = new DefaultFileSystemSnapshotter(fileHasher, new StringInterner(), TestFiles.fileSystem(), fileSystemMirror)
+    def snapshotter = new DefaultFileCollectionSnapshotter(fileHasher, new StringInterner(), fsSnapshotter)
```
Comment (lptr) [inline subprojects/execution/src/integTest/groovy/org/gradle/internal/execution/IncrementalExecutionIntegrationTest.groovy]: This should use the same interner, too, no?

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java [hunk 2]
```diff
@@ -49,7 +48,16 @@
...
 
     /**
-     * Returns snapshots of the roots of a file collection.
+     * Snapshots a directory tree.
+     *
+     * For simplicity this only caches trees without includes/excludes. However, if it is asked
+     * to snapshot a filtered tree, it will try to find a snapshot for the underlying
+     * tree and filter it in memory instead of walking the file system again. This covers the
+     * majority of cases, because all task outputs are put into the cache without filters
+     * before any downstream task uses them.
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java]: Great to have this documented!

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java [hunk 3]
```diff
@@ -49,7 +48,16 @@
...
+     *
+     * If it turns out that a filtered tree has actually not been filtered (i.e. the condition always returned true),
+     * then we cache the result as unfiltered tree.
      */
-    List<FileSystemSnapshot> snapshot(FileCollectionInternal fileCollection);
+    FileSystemSnapshot snapshotDirectoryTree(File root, PatternSet patterns);
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java]: Why did we return a `List` before? :o

File: subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/FileSystemSnapshotBuilderTest.groovy [hunk 4]
```diff
@@ -123,7 +118,8 @@ class FileSystemSnapshotBuilderTest extends Specification {
         def result = builder.build()
 
         then:
-        result == snapshot
+        result instanceof RegularFileSnapshot
+        result.hash == hasher.hash(new File(basePath), snapshot)
```
Comment (lptr) [inline subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/FileSystemSnapshotBuilderTest.groovy]: Why did we need this change?

File: subprojects/execution/src/integTest/groovy/org/gradle/internal/execution/IncrementalExecutionIntegrationTest.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/execution/src/integTest/groovy/org/gradle/internal/execution/IncrementalExecutionIntegrationTest.groovy]: Ah, it's fixed in the next commit, ok.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java (hunk 3 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/FileSystemSnapshotter.java]: The diff is misleading: The old method snapshotted a file collection. Therefore, it returned a list of snapshots. The new method snapshots a single directory tree. So it only returns one snapshot.

File: subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/FileSystemSnapshotBuilderTest.groovy (hunk 4 above)
Comment (wolfs) [reply subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/FileSystemSnapshotBuilderTest.groovy]: `snapshot` is not a `RegularFileSnapshot` any more. The `RegularFileSnapshot` is now created by the `addFile` method. So I can only assert some properties of the result.

**Comment by wolfs** [comment]
@lptr I replied to your comments. PTAL!

**Pull request merged**